File: Japanese_kidney_CKD/code/Normalized_DMP_filter.R
# Normalized_DMP_filter.R

run_normalized_dmp_filter <- function(base_dir) {
  # 設定檔案路徑
  normalization_path <- file.path(base_dir, "csv", "all_beta_normalized.csv")
  dmp_path <- file.path(base_dir, "csv", "DMP_result_TC.csv")
  sample_sheet_path <- file.path(base_dir, "processing_data", "Sample_sheet.csv")
  
  # 讀取資料
  # row.names = 1 相當於 pandas 的 index_col=0
  Normalization <- read.csv(normalization_path, row.names = 1, stringsAsFactors = FALSE)
  DMP <- read.csv(dmp_path, row.names = 1, stringsAsFactors = FALSE)
  Sample_Sheet <- read.csv(sample_sheet_path, stringsAsFactors = FALSE)
  cat("資料讀取完成\n")
  
  # 篩選資料 (取交集)
  common_CpG <- intersect(rownames(Normalization), rownames(DMP))
  cat(sprintf("共有 %d 筆相同的 ID 資料。\n", length(common_CpG)))
  
  # subset 取出共同列，drop = FALSE 確保只有一欄時不會被轉為 Vector
  filtered_Normalization <- Normalization[common_CpG, , drop = FALSE]
  cat(sprintf("篩選完成！共保留了 %d 筆相同的 ID 資料。\n", nrow(filtered_Normalization)))
  
  # 分組 (CKD 與 Normal)
  Sample_Sheet$Sample_Name <- as.character(Sample_Sheet$Sample_Name)
  
  ckd_samples <- Sample_Sheet$Sample_Name[Sample_Sheet$Sample_Group == 'C']
  normal_samples <- Sample_Sheet$Sample_Name[Sample_Sheet$Sample_Group == 'N']
  cat(sprintf("在 Sample Sheet 中找到 %d 個 CKD 樣本，%d 個 Normal 樣本。\n", 
              length(ckd_samples), length(normal_samples)))
  
  # 確保取出的 column name 確實存在於 dataframe 中
  ckd_cols <- intersect(ckd_samples, colnames(filtered_Normalization))
  normal_cols <- intersect(normal_samples, colnames(filtered_Normalization))
  
  EG <- filtered_Normalization[, ckd_cols, drop = FALSE]
  CG <- filtered_Normalization[, normal_cols, drop = FALSE]
  
  write.csv(EG, file.path(base_dir, "csv", "EG_filtered.csv"), row.names = TRUE)
  write.csv(CG, file.path(base_dir, "csv", "CG_filtered.csv"), row.names = TRUE)
  cat("結果已儲存\n")
  
  # 回傳兩個 dataframe 以供後續腳本使用
  return(list(CG = CG, EG = EG))
}

# 模擬 Python 的 if __name__ == "__main__":
if (!interactive() && sys.nframe() == 0) {
  # R 語言沒有直接對應 __file__ 的寫法，這裡以當前工作目錄(getwd)代替 base_dir
  base_dir <- file.path(getwd(), "..")
  result <- run_normalized_dmp_filter(base_dir)
}

File: Japanese_kidney_CKD/code/Normalized_DMP_filter.py
import pandas as pd
import os

class NormalizedDMPFilter:
    def __init__(self, base_dir):
        self.base_dir = base_dir
        self.Normalization_path = os.path.join(self.base_dir, "csv/all_beta_normalized.csv")
        self.DMP_path = os.path.join(self.base_dir, "csv/DMP_result_TC.csv")
        self.Sample_Sheet_path = os.path.join(self.base_dir, "processing_data/Sample_sheet.csv")

        self.Normalization = None
        self.DMP = None
        self.Sample_Sheet = None

        self.filtered_Normalization = None
        self.EG = None
        self.CG = None

    def load_data(self):
        self.Normalization = pd.read_csv(self.Normalization_path, index_col=0, encoding='utf-8')
        self.DMP = pd.read_csv(self.DMP_path, index_col=0, encoding='utf-8')
        self.Sample_Sheet = pd.read_csv(self.Sample_Sheet_path, encoding='utf-8')
        print("資料讀取完成")

    def filter_data(self):
        common_CpG = self.Normalization.index.intersection(self.DMP.index)
        print(f"共有 {len(common_CpG)} 筆相同的 ID 資料。")

        self.filtered_Normalization = self.Normalization.loc[common_CpG]
        self.filtered_Normalization.index.name = 'CpG_ID'
        print(f"篩選完成！共保留了 {len(self.filtered_Normalization)} 筆相同的 ID 資料。")

    def split_groups(self):
        self.Sample_Sheet['Sample_Name'] = self.Sample_Sheet['Sample_Name'].astype(str)

        ckd_samples = self.Sample_Sheet[self.Sample_Sheet['Sample_Group'] == 'C']['Sample_Name'].tolist()
        normal_samples = self.Sample_Sheet[self.Sample_Sheet['Sample_Group'] == 'N']['Sample_Name'].tolist()
        print(f"在 Sample Sheet 中找到 {len(ckd_samples)} 個 CKD 樣本，{len(normal_samples)} 個 Normal 樣本。")

        ckd_cols = [col for col in ckd_samples if col in self.filtered_Normalization.columns]
        normal_cols = [col for col in normal_samples if col in self.filtered_Normalization.columns]

        self.EG = self.filtered_Normalization[ckd_cols]
        self.CG = self.filtered_Normalization[normal_cols]
    
    def save_results(self):
        self.EG.to_csv(os.path.join(self.base_dir, "csv", "EG_filtered.csv"), encoding='utf-8') # 實驗組的輸出檔名
        self.CG.to_csv(os.path.join(self.base_dir, "csv", "CG_filtered.csv"), encoding='utf-8') # 對照組的輸出檔名
        print("結果已儲存")
    
    def run(self):
        self.load_data()
        self.filter_data()
        self.split_groups()
        self.save_results()
        return self.CG, self.EG
    
if __name__ == "__main__":
    base_dir = os.path.dirname(os.path.abspath(__file__)) + '/..'
    filter = NormalizedDMPFilter(base_dir)
    CG, EG = filter.run()
    

#base_dir = os.path.dirname(os.path.abspath(__file__))
#
#Normal = pd.read_csv(os.path.join(base_dir, "csv/all_beta_normalized.csv"), index_col=0, encoding='utf-8')
#DMP = pd.read_csv(os.path.join(base_dir, "csv/DMP_result_TC.csv"), index_col=0, encoding='utf-8')
#
#common_CpG = Normal.index.intersection(DMP.index)
#print(f"共有 {len(common_CpG)} 筆相同的 ID 資料。")
##print("共同欄位名稱:", common_CpG.tolist())
#
#filtered_Normal = Normal.loc[common_CpG]
#filtered_Normal.index.name = 'CpG_ID'
#
##output_path = os.path.join(base_dir, "csv", "Normal_filtered.csv") # 新的輸出檔名
##filtered_Normal.to_csv(output_path, encoding='utf-8')
#
#print(f"篩選完成！共保留了 {len(filtered_Normal)} 筆相同的 ID 資料。")
##print(f"檔案已儲存至：{output_path}")
#
#
##------------------------------分成實驗組與對照組------------------------------#
#
#Sample_Sheet = pd.read_csv(os.path.join(base_dir, "processing_data/Sample_sheet.csv"), encoding='utf-8')
#Sample_Sheet['Sample_Name'] = Sample_Sheet['Sample_Name'].astype(str)
#
#ckd_samples = Sample_Sheet[Sample_Sheet['Sample_Group'] == 'CKD']['Sample_Name'].tolist()
#normal_samples = Sample_Sheet[Sample_Sheet['Sample_Group'] == 'Normal']['Sample_Name'].tolist()
#print(f"在 Sample Sheet 中找到 {len(ckd_samples)} 個 CKD 樣本，{len(normal_samples)} 個 Normal 樣本。")
#
#ckd_cols = [col for col in ckd_samples if col in filtered_Normal.columns]
#normal_cols = [col for col in normal_samples if col in filtered_Normal.columns]
#
#output_EG = filtered_Normal[ckd_cols]
#output_CG = filtered_Normal[normal_cols]
#
#output_EG.to_csv(os.path.join(base_dir, "csv", "EG_filtered.csv"), encoding='utf-8') # 實驗組的輸出檔名
#output_CG.to_csv(os.path.join(base_dir, "csv", "CG_filtered.csv"), encoding='utf-8') # 對照組的輸出檔名